core/vm/analysis.go

```diff
@@ -19,28 +19,99 @@
 
 package vm
 
+const (
+	set2BitsMask = uint16(0b11)
+	set3BitsMask = uint16(0b111)
+	set4BitsMask = uint16(0b1111)
+	set5BitsMask = uint16(0b1_1111)
+	set6BitsMask = uint16(0b11_1111)
+	set7BitsMask = uint16(0b111_1111)
+)
+
 // codeBitmap collects data locations in code.
-func codeBitmap(code []byte) []uint64 {
+func codeBitmap(code []byte) bitvec {
 	// The bitmap is 4 bytes longer than necessary, in case the code
 	// ends with a PUSH32, the algorithm will push zeroes onto the
 	// bitvector outside the bounds of the actual code.
-	bits := make([]uint64, (len(code)+32+63)/64)
+	bits := make(bitvec, len(code)/8+1+4)
+	codeBitmapInternal(code, bits)
+	return bits
+}
+
+// bitvec is a bit vector which maps bytes in a program.
+// An unset bit means the byte is an opcode, a set bit means
+// it's data (i.e. argument of PUSHxx).
+type bitvec []byte
+
+func (bits bitvec) set1(pos uint64) {
+	bits[pos/8] |= 1 << (pos % 8)
+}
+
+func (bits bitvec) setN(flag uint16, pos uint64) {
+	a := flag << (pos % 8)
+	bits[pos/8] |= byte(a)
+	if b := byte(a >> 8); b != 0 {
+		bits[pos/8+1] = b
+	}
+}
 
-	for pc := 0; pc < len(code); {
+func (bits bitvec) set8(pos uint64) {
+	a := byte(0xFF << (pos % 8))
+	bits[pos/8] |= a
+	bits[pos/8+1] = ^a
+}
+
+func (bits bitvec) set16(pos uint64) {
```

Comment:
oh interesting, could use `bitvec` as key representation for bin patricia trie

Reply:
@awskii 
1. `bitvec` i got from geth. but it came there from silkworm (evmone). 
2. bitvec is []byte, but we using []uint64 - and we win in most of cases. we don't use bitvec, but maybe can create []u64 version of bitvec.